File: 468  Symmetric Binary Tree.py
"""***************************  TITLE  ****************************"""
"""468  Symmetric Binary Tree.py"""



"""***************************  DESCRIPTION  ****************************"""
"""
Given a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).
"""



"""***************************  EXAMPLES  ****************************"""
"""
Input: {1,2,2,3,4,4,3}
Output: true
Explanation:
         1
        / \
       2   2
      / \ / \
      3 4 4 3

is a symmetric binary tree.

"""



"""***************************  CODE  ****************************"""
class Solution:
    def isSymmetric(self, root):
        if not root:
            return True
        
        return self.helper(root.left, root.right)
            
    def helper(self, t1, t2):
        if not t1 and not t2:
            return True
            
        if not t1 or not t2:
            return False
            
        return t1.val == t2.val and self.helper(t1.left, t2.right) and \
                                    self.helper(t1.right, t2.left)

File: 468  Symmetric Binary Tree.java
/****************************  TITLE  *****************************/
/*468  Symmetric Binary Tree.java*/



/****************************  DESCRIPTION  *****************************/
/*
Given a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).
*/



/****************************  EXAMPLES  *****************************/
/*
Input: {1,2,2,3,4,4,3}
Output: true
Explanation:
         1
        / \
       2   2
      / \ / \
      3 4 4 3

is a symmetric binary tree.

*/



/****************************  CODE  *****************************/
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the root of binary tree.
     * @return: true if it is a mirror of itself, or false.
     */
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return check(root.left, root.right);
    }
    
    
    public boolean check(TreeNode root1, TreeNode root2) {
        if (root1 == null && root2 == null) return true;
        if (root1 == null || root2 == null) return false;
        if (root1.val != root2.val) return false;
        
        return check(root1.left, root2.right) && check(root1.right, root2.left);
    }
}
